Create logging for constant pool bug

benjamin-confino opened the pull request:
Same PR as #26556 but this one is rebased on integration.

tbitonti committed:
merge in tomb's logging

tbitonti committed:
Trace instrumentation adjustements.

tbitonti committed:
Add logging to JBOSS WELD reflection cache.

tbitonti committed:
Trace adjustments.  Add trace to weld4.

tbitonti committed:
Put back jacc_fat test classes.
Enable weld hotspot logging using system properties.
Revert unnecessary liberty trace instrumentation update.

benjamin-confino committed:
create property for logging

benjamin-confino committed:
fix errors

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino commented:
#build

tevans78 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/HotspotReflectionCache.java:
```diff
@@ -0,0 +1,157 @@
+
+        try {
+            prop = getSystemProperty("jboss.hotspot.trace.enabled");
+        } catch (Exception ignored) {}
```
FFDCIgnore was for Throwable but you caught Exception. Why do you want to ignore the exception at all anyway?

benjamin-confino replied:
To keep this from affecting anything other than the servers we're adding this unique logging too. 

Doesn't FFDCIgnore also deal with subtypes?

tevans78 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/HotspotReflectionCache.java:
```diff
@@ -0,0 +1,157 @@
+    boolean sentWarning = false;
+
+    @Override
+    @FFDCIgnore(Throwable.class)
```
```suggestion
    @FFDCIgnore(Exception.class)
```

tevans78 replied:
or remove it

tevans78 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/HotspotReflectionCache.java:
```diff
@@ -0,0 +1,157 @@
+/*
```
Please add clear references to where the original source was copied from and clearly mark what you have changed.

benjamin-confino commented:
#build

una-tapa reviewed (dismissed):
Looks good to me. Thanks for explaining Ben.

jhanders34 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/HotspotReflectionCache.java:
```diff
@@ -0,0 +1,161 @@
+        String prop = "";
+
+        try {
+            prop = getSystemProperty("jboss.hotspot.trace.enabled");
```
Shouldn't this be done statically instead of inline to the method?  Or are you expecting that people are going to call System.setProperty() to set it?

jhanders34 replied:
I see it is now done statically, but the code is still here.  The prop = "" and the try / catch can be removed I believe.

jhanders34 commented on dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/AbstractInstrumentation.java:
```diff
@@ -368,7 +367,7 @@ public void instrumentZipFile(File zf) throws IOException {
 
                 if (oldEntry.getName().endsWith(".class")) {
-                    transformedClass = transform(zis);
+                    transformedClass = transform(oldEntry.getName(), zis);
```
This is not the class name.  It is the name of the file.  There is inconsistency between what is passed to the transform method.

jhanders34 commented on dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/AbstractInstrumentation.java:
```diff
@@ -327,7 +326,7 @@ private boolean isStreamForClass(InputStream inputStream, String classInternalNa
     public void instrumentClassFile(File classfile) throws IOException {
         FileInputStream fis = new FileInputStream(classfile);
-        byte[] bytes = transform(fis);
+        byte[] bytes = transform(classfile.getPath(), fis);
```
Same here.  This is not the class name.  It is a file path.

jhanders34 commented on dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/StaticTraceInstrumentation.java:
```diff
@@ -119,111 +157,179 @@ public void setInstrumentationOptions(InstrumentationOptions instrumentationOpti
+    // removed.  The trace configuration visitor is a read-only
+    // visitor.  There is no need to rewrite the class bytes.
+
+    private ClassInfo readConfig_new(byte[] classBytes) throws IOException {
```
This method isn't used.  Should we be using it instead of readConfig?

jhanders34 commented on dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java:
```diff
@@ -368,6 +393,67 @@ public static byte[] transform(byte[] bytes, boolean skipIfNotPreprocessed) thro
+        
+        LibertyTracingClassAdapter tracingClassAdapter = new LibertyTracingClassAdapter(classVisitor, skipIfNotPreprocessed);
+        try {
+            classReader.accept(tracingClassAdapter, skipDebugData ? ClassReader.SKIP_DEBUG : 0);
```
No longer is the code handling the case with the compute max and the throw computer frames.  Is that intentional?  Comment from the previous code:

```
		// Occasionally, frames must be computed.  This is not known until
		// the class is visited, as the circumstances depend on encountering
		// variable manipulating byte-codes before invoking the superclass
		// initializer.
		//
		// The (hopefully rare) occasions are communicated by throwing
    	// 'ComputeRequiredException'.
                
        // First try: Use COMPUTE_MAXS and THROW_COMPUTE_FRAMES in case
        // the flag must be reset.
```

benjamin-confino committed:
code review requested changes

tbitonti commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/HotspotReflectionCache.java:
The constructor has stale property retrieval code.

tbitonti commented on dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/DynamicTraceInstrumentation.java:
Ah, we might need to be more careful.  That 'startsWith' is not adequate in several cases:
The entry name of a multi-version class will have extra path elements.
The entry name of WEB-INF/classes classes will have extra path elements.
A path obtained from a File will depend on particulars of how the file was obtained.

tbitonti committed:
Remove incorrect renames of 'path' to 'classPath'.
Split 'isLoggable' to 'isLoggablePath' and 'isLoggableClasspath', which is required because of the dual usage.
Handle windows slashes correctly in 'isLoggablePath'.
More completely document 'path'.
Re-merge LibertyRuntimeTransformer to put back in the frame computation fix.
More completely document the liberty runtime transformer steps.

tbitonti committed:
transform API corrections

benjamin-confino committed:
Merge pull request #3 from tbitonti/ti52

Ti52

tbitonti commented:
Successfully built here:
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_vNIdkJqXEe6q84kI3KZ5lw&tab=com.ibm.team.build.web.ui.internal.editors.result.overview.OverviewPage

jhanders34 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/FileLogger.java:
```diff
@@ -0,0 +1,707 @@
+    public static final boolean enabled;
+
+    static {
+        String enabledValue = System.getProperty(ENABLED_PROPERTY_NAME);
```
Isn't doPriv needed here?

tbitonti replied:
Added.

jhanders34 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/FileLogger.java:
```diff
@@ -0,0 +1,707 @@
+        return System.currentTimeMillis();
+    }
+
+    private static final SimpleDateFormat formatter =
```
SimpleDateFormat is not thread safe.  Better to use a DateTimeFormatter.  You may need to change a few of the letters.  I changed several SimpleDateFormat's in liberty to DateTimeFormatter objects.  If you need any help let me know.

tbitonti replied:
Added sync.

jhanders34 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/FileLogger.java:
```diff
@@ -0,0 +1,707 @@
+            Properties useProperties = new Properties();
+
+            for ( String propertyName : PROPERTY_NAMES ) {
+                String propertyValue = System.getProperty(propertyName);
```
This also should be using doPriv to get System properties

tbitonti replied:
Added doPriv.

jhanders34 commented on dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/DynamicTraceInstrumentation.java:
```diff
@@ -150,22 +167,23 @@ class Transformer extends StaticTraceInstrumentation implements ClassFileTransfo
+        String methodName = "transform";
+        
         // Don't modify our own package
         if (className.startsWith(Transformer.class.getPackage().getName().replaceAll("\\.", "/"))) {
```
I thought you were looking at changing this to contains instead of startsWith?  If it is a full URI path, it will include additional content than just the relative path to the file.  It may be better to strip off the .class and do an endsWith?

Also if it is a className instead of a path, why are we convert the package name to have slashes instead of periods?

tbitonti replied:
This occurrence is really a class name.  Entry is from 'com.ibm.ws.classloading.internal.AppClassLoader.doTransformation(String, byte[], ClassFileTransformer)'.
The prior update to change this to 'path' was incorrect.  This method should not have been updated.

jhanders34 replied:
In reality what is passed from AppClassLoader.doTransformation is the wrong thing since it is the period separated class name instead of slash name, but it won't cause any problems since application classes will not start with the packages we are filtering on.  It will cause issues with the debug though since we are passing the className as a path when passing it with the InputStream transform method instead of the byte[] transform method.  You can't change to use the byte[] one since that will be incorrect in the re-transform path from the JDK.  As such it is probably best to change the doTransform logic to pass in the slash separated name.  Could pass in the resourceName, but that does include the `.class` at the end of the String.

jhanders34 replied:
Actually will AppClassLoader.doTransformation() even use this Transformer?  I didn't think we did trace transformation for application classes.  I thought we only did it for run time classes.  If so this isn't a problem for this PR, but is a problem still for the transformers that are called by AppClassLoader and can be done under a separate PR.

jhanders34 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/FileLogger.java:
```diff
@@ -0,0 +1,707 @@
+
+    public static String getFormattedTime() {
+        long currentMs = getTime();
+        if ((currentMs - current.getTime()) > 10)  {
+            current.setTime(currentMs);
+            currentFormatted = formatter.format(current);
+        }
```
In a multi threaded environment this isn't going to be thread safe either.

tbitonti replied:
Added sync.

tbitonti committed:
Added doPrivileged and synchronization per review.

tbitonti committed:
Merge branch 'integration' into ti52

jhanders34 commented on dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/FileLogger.java:
```diff
@@ -89,11 +92,28 @@ public static byte[] read(InputStream inputStream) throws IOException {
+            public String run() {
+                return getSystemProperty(ENABLED_PROPERTY_NAME);
+            }
+        });
```
```suggestion
        String enabledValue = getSystemProperty(ENABLED_PROPERTY_NAME);
```

jhanders34 commented on dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/StaticTraceTestTransformer.java:
```diff
@@ -29,11 +29,11 @@ public class StaticTraceTestTransformer extends LibertyTracePreprocessInstrument
     	
         try {
-            return transform(new ByteArrayInputStream(classfileBuffer));
+            return transform(className, new ByteArrayInputStream(classfileBuffer));
```
Since this isn't testing the debug function, I can live with it, but it really should convert the periods in the class name with slashes when passing to this method.

tbitonti replied:
Put in followup issue #27420.

tbitonti committed:
Update dev/com.ibm.ws.org.jboss.weld/src/org/jboss/weld/resources/FileLogger.java

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

jhanders34 commented:
#libby
